tests/host1x/gr2d-context.c

```diff
@@ -0,0 +1,306 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "host1x.h"
+#include "host1x-private.h"
+
+#define MIN(a, b) ((a) < (b) ? (a) : (b))
+
+#define CTX_NB 11
+
+struct ctx2d {
+	struct host1x *host1x;
+	struct host1x_gr2d *gr2d;
+};
+
+static struct ctx2d *create_context()
+{
+	struct ctx2d *ctx = calloc(1, sizeof(struct ctx2d));
+
+	ctx->host1x = host1x_open(true, -1);
+	if (!ctx->host1x) {
+		fprintf(stderr, "host1x_open() failed\n");
+		abort();
+	}
+
+	ctx->gr2d = host1x_get_gr2d(ctx->host1x);
+	if (!ctx->gr2d) {
+		fprintf(stderr, "host1x_get_gr2d() failed\n");
+		abort();
+	}
+
+	return ctx;
+}
+
+static void prepare_context(struct ctx2d *ctx,
+			    struct host1x_pixelbuffer *target,
+			    unsigned int width, unsigned int height,
+			    uint32_t color)
+{
+	struct host1x_syncpt *syncpt = &ctx->gr2d->client->syncpts[0];
+	struct host1x_pixelbuffer *src;
+	struct host1x_pushbuf *pb;
+	struct host1x_job *job;
+	struct host1x_bo *dst;
+	uint32_t handle;
+	uint32_t fence;
+	int err;
+
+	HOST1X_BO_EXPORT(target->bo, &handle);
+	dst = HOST1X_BO_IMPORT(ctx->host1x, handle);
+
+	src = host1x_pixelbuffer_create(ctx->host1x,
+					width, height, width * 4,
+					PIX_BUF_FMT_RGBA8888,
+					PIX_BUF_LAYOUT_LINEAR);
+	if (!src)
+		abort();
+
+	err = host1x_gr2d_clear(ctx->gr2d, src, color);
+	if (err < 0)
+		abort();
+
+	job = HOST1X_JOB_CREATE(syncpt->id, 1);
+	if (!job)
+		abort();
+
+	pb = HOST1X_JOB_APPEND(job, ctx->gr2d->commands, 0);
+	if (!pb)
+		abort();
+
+	host1x_pushbuf_push(pb, HOST1X_OPCODE_SETCL(0, 0x51, 0));
+
+	host1x_pushbuf_push(pb, HOST1X_OPCODE_MASK(0x009, 0x9));
+	host1x_pushbuf_push(pb, 0x0000003a); /* trigger */
+	host1x_pushbuf_push(pb, 0x00000000); /* cmdsel */
+
+	host1x_pushbuf_push(pb, HOST1X_OPCODE_MASK(0x01e, 0x7));
+	host1x_pushbuf_push(pb, 0x00000000); /* controlsecond */
+	/*
+	 * [20:20] source color depth (0: mono, 1: same)
+	 * [17:16] destination color depth (0: 8 bpp, 1: 16 bpp, 2: 32 bpp)
+	 */
+	host1x_pushbuf_push(pb, /* controlmain */
+			1 << 20 |
+			(PIX_BUF_FORMAT_BYTES(target->format) >> 1) << 16 |
+			0 << 14 | 0 << 10 | 0 << 9);
+	host1x_pushbuf_push(pb, 0x000000cc); /* ropfade */
+
+	host1x_pushbuf_push(pb, HOST1X_OPCODE_NONINCR(0x046, 1));
+	/*
+	 * [20:20] destination write tile mode (0: linear, 1: tiled)
+	 * [ 0: 0] tile mode Y/RGB (0: linear, 1: tiled)
+	 */
+	host1x_pushbuf_push(pb, 0 << 20 | 0); /* tilemode */
+
+	host1x_pushbuf_push(pb, HOST1X_OPCODE_MASK(0x02b, 0x6149));
+	HOST1X_PUSHBUF_RELOCATE(pb, dst, target->bo->offset, 0);
```

Comment:
Since the rest of the code calls abort in case of trouble, perhaps just check if HOST1X_BO_EXPORT failed, and abort there also?